src/kms.rs

```diff
@@ -0,0 +1,329 @@
+                }
+            })
+            .or_else(|| handles.connectors.first().copied())
+            .swbuf_err("failed to find a valid connector")?;
+
+        Ok(Self {
+            crtc: display
+                .get_crtc(crtc)
+                .swbuf_err("failed to get CRTC info")?,
+            connector,
+            display,
+            buffer: None,
+        })
+    }
+
+    /// Resize the internal buffer to the given size.
+    pub(crate) fn resize(
+        &mut self,
+        width: NonZeroU32,
+        height: NonZeroU32,
+    ) -> Result<(), SoftBufferError> {
+        // Don't resize if we don't have to.
+        if let Some(buffer) = &self.buffer {
+            let (buffer_width, buffer_height) = buffer.size();
+            if buffer_width == width && buffer_height == height {
+                return Ok(());
+            }
+        }
+
+        // Create a new buffer.
+        let buffer = BufferSet::new(&self.display, width, height)?;
+
+        // Set the framebuffer in the CRTC info.
+        // TODO: This requires root access, find a way that doesn't!
+        self.display
+            .set_crtc(
+                self.crtc.handle(),
+                Some(buffer.fb),
+                self.crtc.position(),
+                &[self.connector],
+                self.crtc.mode(),
+            )
+            .swbuf_err("failed to set CRTC")?;
+
+        self.buffer = Some(buffer);
+
+        Ok(())
+    }
+
+    /// Fetch the buffer from the window.
+    pub(crate) fn fetch(&mut self) -> Result<Vec<u32>, SoftBufferError> {
+        // TODO: Implement this!
+        Err(SoftBufferError::Unimplemented)
+    }
+
+    /// Get a mutable reference to the buffer.
+    pub(crate) fn buffer_mut(&mut self) -> Result<BufferImpl<'_>, SoftBufferError> {
+        // Map the dumb buffer.
+        let set = self
+            .buffer
+            .as_mut()
+            .expect("Must set size of surface before calling `buffer_mut()`");
+        let size = set.size();
+        let mapping = self
+            .display
+            .map_dumb_buffer(&mut set.db)
+            .swbuf_err("failed to map dumb buffer")?;
+
+        Ok(BufferImpl {
+            mapping,
+            size,
+            fb: set.fb,
+            display: &self.display,
+            zeroes: &set.zeroes,
+        })
+    }
+}
+
+impl Drop for KmsImpl {
+    fn drop(&mut self) {
+        // Map the CRTC to the information that was there before.
+        self.display
+            .set_crtc(
+                self.crtc.handle(),
+                self.crtc.framebuffer(),
+                self.crtc.position(),
+                &[self.connector],
+                self.crtc.mode(),
+            )
+            .ok();
+    }
+}
+
+impl BufferImpl<'_> {
+    #[inline]
+    pub fn pixels(&self) -> &[u32] {
+        // drm-rs doesn't let us have the immutable reference... so just use a bunch of zeroes.
+        // TODO: There has to be a better way of doing this!
```

Comment:
Well, the `BorrowStack` hack would work here.

But it looks like `DumbMapping` is already just a wrapper around a `&'a mut [u8]`, so could just open a PR for drm-rs to implement `AsRef<[u8]>` in addition to the `AsMut<[u8]>` implementation.